### PR TITLE
[GR-41977] Adapt to JVMCI API changes.

### PR DESCRIPTION
**Pull request by graalvmbot**

### CODE REVIEW

File: compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java [hunk 1]
```diff
@@ -150,52 +122,23 @@ public Class<?> originalClass(ResolvedJavaType type) {
...
-
-        if (hotSpotJDKReflectionGetMethod == null) {
-            return null;
-        }
-        try {
-            return (Executable) hotSpotJDKReflectionGetMethod.invoke(null, method);
-        } catch (ReflectiveOperationException ex) {
-            throw rethrow(ex.getCause());
-        }
+        return runtime().getMirror(method);
```
Comment (zakkak) [inline compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java:134]: This results in:

```
/home/runner/work/mandrel/mandrel/mandrel/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java:134: error: incompatible types: ResolvedJavaMethod cannot be converted to ResolvedJavaType
        return runtime().getMirror(method);
```

When building with JDK 17 (both OpenJDK and LabsJDK).

See https://github.com/graalvm/mandrel/actions/runs/3499526212/jobs/5861191268#step:8:445 and https://github.com/graalvm/mandrel/actions/runs/3499602031/jobs/5861349246#step:7:330, respectively.

File: compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java [hunk 2]
```diff
@@ -150,52 +122,23 @@ public Class<?> originalClass(ResolvedJavaType type) {
...
-            throw rethrow(ex.getCause());
-        }
-    }
+        Objects.requireNonNull(field);
+        GraalError.guarantee(field instanceof HotSpotResolvedJavaField, "Unexpected implementation class: %s", field.getClass());
 
-    @SuppressWarnings({"unchecked"})
-    private static <E extends Throwable> RuntimeException rethrow(Throwable ex) throws E {
-        throw (E) ex;
+        return runtime().getMirror(field);
```
Comment (zakkak) [inline compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java:142]: This results in:

```
/home/runner/work/mandrel/mandrel/mandrel/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/meta/HotSpotSnippetReflectionProvider.java:142: error: incompatible types: ResolvedJavaField cannot be converted to ResolvedJavaType
        return runtime().getMirror(field);
```

When building with JDK 17 (both OpenJDK and LabsJDK).

See https://github.com/graalvm/mandrel/actions/runs/3499526212/jobs/5861191268#step:8:445 and https://github.com/graalvm/mandrel/actions/runs/3499602031/jobs/5861349246#step:7:330, respectively.

**Review by zakkak — CHANGES_REQUESTED** [review]
Hello @dougxc @christianwimmer, are there any plans to backport the corresponding JVMCI changes to OpenJDK 17? If not, could we please use a more flexible approach to avoid breaking compatibility with JDK 17?

**Comment by jerboaa** [comment]
Thanks for the heads-up. Looking into it.

**Comment by christianwimmer** [comment]
The current plan is that nothing from the current master branch will be released anymore based on JDK 17. We are switching to "release new GraalVM versions only with the latest Java version", i.e., GraalVM 23.0 will be based (and only based) on Java 20.

But in case that plan changes, or you want to release something based on JDK 17, there are only a few small OpenJDK issues that you need to backport into OpenJDK 17:
* https://bugs.openjdk.org/browse/JDK-8296958
* https://bugs.openjdk.org/browse/JDK-8296961
* https://bugs.openjdk.org/browse/JDK-8296960
* https://bugs.openjdk.org/browse/JDK-8296956
* https://bugs.openjdk.org/browse/JDK-8296967

**Comment by jerboaa** [comment]
> * https://bugs.openjdk.org/browse/JDK-8296958
> * https://bugs.openjdk.org/browse/JDK-8296961
> * https://bugs.openjdk.org/browse/JDK-8296960
> * https://bugs.openjdk.org/browse/JDK-8296956
> * https://bugs.openjdk.org/browse/JDK-8296967

FWIW, all of these have been either backported to the `jdk17u-dev` tree already or are in progress of getting integrated. Some waiting for approvals.

**Comment by zakkak** [comment]
@jerboaa it looks like all the listed patches have been backported and will be part of 17.0.6. Is that correct? Can this PR be integrated?

FWIW Mandrel seems to build fine with this patch and OpenJDK 17.0.6-beta+6-202212140702 (see https://github.com/graalvm/mandrel/actions/runs/3702797693/jobs/6273484993)

**Comment by jerboaa** [comment]
> @jerboaa it looks like all the listed patches have been backported and will be part of 17.0.6. Is that correct?

Correct.

**Comment by fniephaus** [comment]
Does this still need to be backported to JDK 17?

**Comment by jerboaa** [comment]
> Does this still need to be backported to JDK 17?

AFAIK, all changes are in OpenJDK `17.0.6` and better. So no.

**Comment by fniephaus** [comment]
> AFAIK, all changes are in OpenJDK 17.0.6 and better. So no.

Ok good, I thought this PR is about adapting JVMCI API changes in GraalVM and that needs to be backported to work with OpenJDK 17.0.6 or later.

**Pull request closed without merging**